### DISCUSSION TITLE
Support for tensorflow and fortran 90?

### DISCUSSION THREAD

**Discussion by xushanthu-2014** [ylikx/forpy#34] [Q&A]
Currently I am running an atmospheric model which was written by Fortran 90. In some subroutines I made some innovations using python (mainly tensorflow). Now I need to implement this python code into Fortran, so is there any wrapper to do this? just calling python within a Fortran subroutine? 
I knew that your ```forpy``` could do this. But I am suspecting could it help me to import tensorflow into Fortran?

Here is the pseudo code of my basic thoughts:
```
SUBROUTINE demo(INPUTS, OUTPUTS)
! Definition of INPUTS
REAL, DIMENSION(:,:,:,:), INTENT(IN) :: PINPUTS
REAL, DIMENSION(:,:,:,:), INTENT(IN) :: POUTPUTS

PINPUTS = INPUTS

! using some python wrappers
POUTPUTS = python_wrapper(PINPUTS)

OUTPUTS = POUTPUTS
END SUBROUTINE demo
```

And in the python script "python_wrapper.py", I need to do:
```
import tensorflow as tf

# read saved model "model_saved";
# get the variables "OUTPUTS" passed from demo.f90;

prediction = model_saved.predict(OUTPUTS)

# then pass this variable "prediction" back to Fortran subroutine demo.f90;
# so that the "prediction" will be used for following calculation of the atmospheric model
```

And with ```forpy```, can we simply the process and do the following in the ```demo.f90```?

```
SUBROUTINE demo_sim(INPUTS, OUTPUTS)
! Definition of INPUTS
REAL, DIMENSION(:,:,:,:), INTENT(IN) :: PINPUTS
REAL, DIMENSION(:,:,:,:), INTENT(IN) :: POUTPUTS

PINPUTS = INPUTS

! using forpy to import the tf model saved in my computer
python_model = forpy(./saved_model)
POUTPUTS = python_wrapper(PINPUTS)

OUTPUTS = POUTPUTS
END SUBROUTINE demo_sim
```

Or just in the ```demo.f90```, can ```forpy``` do the wrapper things?

One more thing, the subroutine ```demo``` or. ```demo_sim``` needs to be called in other subroutine (for example, ```outer.f90```) of the atmospheric model. So how can I use this subroutine? I see that in your readme file I need to write something like 
```
# Python 3.8 and higher
gfortran -c forpy_mod.F90
gfortran intro_to_forpy.F90 forpy_mod.o `python3-config --ldflags --embed
```
So do I still need to write the above commands in the ```outer.f90```?

Thanks a lot!

**Comment by ylikx** [comment]
Yes, it should be possible to do this with forpy. I think writing a python module to create a simple interface to tensorflow that is suited to your problem is a good approach. While you could import tensorflow and setup all the tensorflow stuff in Fortran in forpy, this would need a lot of boilerplate code.

I have not used tensorflow with forpy myself, but other people have:
https://github.com/ylikx/forpy/issues/21#issuecomment-604867561

It should also not be a problem to use the subroutine `demo` in several places in your code, I'd just recommend to do the initialisation of forpy (`forpy_initialize`) outside of the subroutine, e.g. in the beginning of your program.

**Reply by xushanthu-2014** [reply]
Hi @ylikx, I cannot import tensorflow or keras with forpy. I will reopen an issue since you move this issue to the discussion. thanks!
